### PR TITLE
use RRect to draw avatar check on chip

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/24773

Previously the darkened layer was drawn incorrectly when combined with the default avatar. I believe this was due to clipping changes. Nevertheless, a clip isn't necessary since we know the shape of the avatar.

We can support this by adding a (please help me name this better) "AvatarShapeDescriber", which given the rect of the avatar, returns a Path which we use to paint the darkened layer. I explored other options such as a widget or painter, but I felt that it exposed too many of the lower-level bits of the chip.

```
typedef AvatarShapeDescriber = Path Function(Rect avatarRect);

Path _kDefaultAvatarPath(Rect avatarRect) {
  final Radius radius = Radius.circular(avatarRect.shortestSide / 2);
  return Path()..addRRect(RRect.fromRectAndRadius(avatarRect, radius));
}
```

Now the default implementation and a circle avatar look correct:

This can also be customized to suit any shape without using a clip, for example a square with rounded corners:

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -2357,7 +2357,8 @@ class _RenderChip extends RenderBox {
         final Paint darkenPaint = Paint()
           ..color = selectionScrimTween.evaluate(checkmarkAnimation)
           ..blendMode = BlendMode.srcATop;
-        context.canvas.drawRect(avatarRect, darkenPaint);
+        final double shortestSide = math.min(avatarRect.width, avatarRect.height);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: You can use `avatarRect.shortestSide`.

File: packages/flutter/lib/src/material/chip.dart [hunk 2]
```diff
@@ -2357,7 +2357,8 @@ class _RenderChip extends RenderBox {
...
+        context.canvas.drawRRect(RRect.fromRectAndRadius(avatarRect, Radius.circular(shortestSide / 2)), darkenPaint);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: I assume this code will work as expected if the avatar's width and height are the same and it paints within a circle.

What if avatarRect is a rectangle and the avatar paints in a circle that's centered within its rect. Will this RRect line up with the avatar?

**Comment by jonahwilliams** [comment]
Making this configurable is a bit of a head scratcher

**Comment by HansMuller** [comment]
I think this solution is on the right path (so to speak). 

The avatar's shape could be defined with a `ShapeBorder` instead of a special callback and the default shape could be defined with a `CircleBorder`.

**Review by HansMuller — APPROVED** [review]
LGTM modulo the name change, doc tweaks.

**Pull request merged**